### DISCUSSION TITLE
Do we need RemixComponents? The backend-frontend components.

### DISCUSSION THREAD

**Discussion by luispuig** [remix-run/remix#2782] [Proposals]
I will try to explain it in a use case:

![use-case](https://user-images.githubusercontent.com/24297787/163234441-acea98bc-38ad-4d6e-baab-be6b567734ca.gif)

Let's assume we need the route to handle "many" CRUD operations. With only one CRUD it is pretty straightforward using the loader/useLoaderData and the action. 

What are the challenges of having many CRUD operations on the same route?
- The data for all the elements being handled in the same loader
- All the actions for All the crud will potentially be handled in the same action. (We have useFetcher but useLoaderData cannot be pointed to another action)
- If we split the UI into different fronted-components we will probably not have the form and the action in the same file.

Of course, I'd love to create some component that I call `Box` that handles the show-data/show-form behavior across the routes.

**How do you think it is a good Remix strategy to implement it?**

# Possible solution: The RemixComponent

One possible solution is the possibility of creating RemixComponent, no frontend, no backend, a frontend-backend component. It is not much different to a Route. The RemixComponent could have a loader, action, and then be rendered in a route. 

In a very old framework I created for myself using Zend Framework, the box component way of use was like this:

```html
<h1 id='section_title'>Empresas</h1>
<span class='full-form'>
	<span class='row-fluid'>
			<span class='span5'>
				<?php echo $this->datatable('dt-images', 'new-after', array('alias' => 'empresas', 'section' => 'empresas') ) ?>
			</span>
			<span class='span7'>
				<?php echo $this->box('seo'); ?>
			</span>
	</span>
</span>

<span class='row-fluid'>
		<span class='span6'>
			<?php echo $this->box('texto'); ?>
		</span>
</span>
```
Files with the HTML for the show-data/edit-data
<img width="275" alt="image" src="https://user-images.githubusercontent.com/24297787/163236342-355120fb-41aa-4f2f-b80c-a03e5ec1619a.png">

Each box has associated a dedicated "controller" for the server-side.

# Disclaimer

Of course, I am interested in my personal use case strategy, but my point is to use it for proposing the RemixComponent or discussing a good way to handle similar situations. I can think of many more use-cases kind of similar.

# Extra

Will it allow the use of directly RemixComponets from an npm package with all the logic for the backend and the frontend? I guess they can be configured somehow.

**Comment by kiliman** [comment]
I think stuff like this are good examples of external packages. Remix Utils is a prime example. A lot of ideas were implemented in this library. Some of them eventually made it into Remix core.

So definitely create a package for this, and I'm sure others that have similar needs will check it out.

Good luck!

**Reply by luispuig** [reply]
I'll be happy to start working on the package. I got some possible ideas. Would anybody like to discuss different options?

**Comment by luispuig** [comment]
I started a little experiment: remix-tracks. What do you think about it? https://github.com/sergiodxa/remix-utils/discussions/88 / https://github.com/luispuig/remix-tracks
